### PR TITLE
simplify API for use in backend

### LINKED ISSUE #81: Change client initalization to be a class instantiation.
Dealing with a promise-based client initialization seems very out of place so I don't believe we should be using it.

### PR DESCRIPTION
**Pull request by isaaccormack**
Now that we have a clearer picture of how the package will be used on the backend, the API can be changed to support only what is needed. This PR does that.

In this PR:
- refactor API to 4 simple stateless functions
- update API tests

This closes #81

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -10,156 +10,71 @@ import {
...
-};
-
-export const UVicCourseScraper = async () => {
-  // upon initialization, we download the main Kuali courses JSON file.
-  const kuali = await got(COURSES_URL).json<KualiCourseCatalog[]>();
-
-  // a map to map human readable subjectCode to a pid required for requests
-  const pidMap = subjectCodeToPidMapper(kuali);
-
+export const UVicCourseScraper = () => {
```
Comment (keithradford) [inline src/index.ts:17]: This was originally created as a function as appose to a class in order to have an async "constructor". Now that we don't need that, maybe it makes sense to turn this into a class again.

File: src/index.ts [hunk 2]
```diff
@@ -10,156 +10,71 @@ import {
...
    */
-  // TODO: Implement functionality for term parameter.
-  // This will be a result of adding logic which uses the proper Kuali/BAN1P link
-  const getCourseDetails = async (
-    subject: string,
-    code: string
-    // TODO: term: string = getCurrentTerm()
-  ): Promise<KualiCourseItem> => {
-    return fetchCourseDetails(pidMap)(subject, code);
+  const getSeats = async (term: string, crn: string): Promise<DetailedClassInformation> => {
```
Comment (keithradford) [inline src/index.ts:69]: Should this be called `getSeats` if it's not all it's returning? It's definitely the main purpose so might not be that big of a deal...

File: src/__tests__/index.test.ts [hunk 3]
```diff
@@ -69,8 +89,8 @@ describe('call getCourseSections', () => {
         .reply(200, detailsResponse);
```
Comment (keithradford) [inline src/__tests__/index.test.ts:89]: This for loop of nocks shouldn't be needed anymore since this function doesn't get section details anymore.

File: src/index.ts (hunk 1 above)
Comment (isaaccormack) [reply src/index.ts:17]: Agreed, making it a class with static methods will make it easy to stub the methods when unit testing functions which use this package (ie. on the backend)

File: src/__tests__/index.test.ts (hunk 3 above)
Comment (isaaccormack) [reply src/__tests__/index.test.ts]: thanks for catching that, went through the rest of these tests in most recent commit to make sure I didn't nock anything else not needed

File: src/index.ts (hunk 2 above)
Comment (isaaccormack) [reply src/index.ts]: I changed it to getSectionSeats to make it a little more descriptive. The only other things its returning besides seat info is requirements, which may or may not exist. I think it's better to call this getSectionSeats than something more generic because 1. the intent of getSectionsSeats is clear, 2. I predict this will be used to get seating info 90% of the time. If that changes, we can consider a name change.

File: src/index.ts [hunk 4]
```diff
@@ -10,156 +10,65 @@ import {
...
    *
-   * @param crn ie. '12345', '20001'
-   * @param term i.e. '202009', '202101'
-   * @return {Promise<DetailedClassInformation>}
+   * @param pid ie. ByS23Pp7E
    */
-  const getSeats = async (crn: string, term: string): Promise<DetailedClassInformation> => {
-    const { seats, waitListSeats } = await fetchSectionDetails(crn, term);
-    return { seats, waitListSeats };
+  public static async getCourseDetails(pid: string): Promise<KualiCourseItem> {
```
Comment (aomi) [inline src/index.ts:33]: We should either allow course details being queried by a human-friendly way OR provide this as a separate method. Perhaps we call this `getCourseDetailsByPid`. Also, we should provide a method to get this pid mapping to the client if they wish to use it. In our use-case, we need the PID mapping too so.

File: src/index.ts [hunk 5]
```diff
@@ -10,156 +10,65 @@ import {
...
- * Generates a Map that maps a subject and code to a pid used internally within Kuali.
- * @param kuali
- */
-const subjectCodeToPidMapper = (kuali: KualiCourseCatalog[]) => {
-  const dict: Map<string, string> = new Map();
-  kuali.forEach((v) => {
-    dict.set(v.__catalogCourseId, v.pid);
-  });
-  return dict;
-};
```
Comment (aomi) [inline src/index.ts:27]: Move this as a static helper method for the user to get pid mappings?

File: src/dev/utils.ts [hunk 6]
```diff
@@ -1,7 +1,7 @@
 import async from 'async';
 import ProgressBar from 'progress';
 
-export const forEachHelper = async <T>(items: T[], asyncfn: (item: T) => void, rateLimit: number) => {
+export const forEachHelper = async <T>(items: T[], asyncfn: (item: T) => void, rateLimit: number): Promise<void> => {
```
Comment (aomi) [inline src/dev/utils.ts:4]: We shouldn't be doing random fixes which aren't in the scope of the issue or PR in the future

File: package.json [hunk 7]
```diff
@@ -1,9 +1,9 @@
 {
-  "name": "@vikelabs/uvic-course-scraper",
-  "version": "0.1.0",
+  "name": "@isaaccormack/uvic-course-scraper",
+  "version": "0.1.4",
   "description": "UVic Course Scrapper",
-  "main": "dist/index.js",
-  "types": "dist/index.d.ts",
+  "main": "dist/src/index.js",
+  "types": "dist/src/index.d.ts",
```
Comment (aomi) [inline package.json:6]: Remember to revert these changes.

File: src/index.ts (hunk 4 above)
Comment (isaaccormack) [reply src/index.ts:33]: Ok. Would it be reasonable for the human-friendly getCourseDetails(term, subject, code) to query getAllCourses() on its first call to create the PID map, then save it internally? Ie. so we don't have an async constructor like before?

File: src/dev/utils.ts (hunk 6 above)
Comment (isaaccormack) [reply src/dev/utils.ts:4]: good point

File: package.json (hunk 7 above)
Comment (isaaccormack) [reply package.json:6]: thanks!

File: src/index.ts [hunk 8]
```diff
@@ -10,156 +10,91 @@ import {
...
-    }));
-  };
+  // TODO: support get course details by term
+  public async getCourseDetails(subject: string, code: string): Promise<KualiCourseItem> {
+    if (UVicCourseScraper.subjectCodeToPidMap.size == 0) {
+      const courseCatalog = await UVicCourseScraper.getAllCourses();
+      UVicCourseScraper.subjectCodeToPidMapper(courseCatalog);
+    }
+
+    const pid = UVicCourseScraper.subjectCodeToPidMap.get(subject + code) as string;
```
Comment (aomi) [inline src/index.ts:48]: Don't we want to uppercase here too?

**Review by aomi — APPROVED** [review]
`export * from './types';` add this to `index.ts` so can import types from the index then I think this is golden.

File: src/index.ts (hunk 8 above)
Comment (isaaccormack) [reply src/index.ts:48]: Yes, thanks. Fun fact - code is not case sensitive from my testing. ie. 485a and 485A both work.

Comment (aomi) [reply src/index.ts]: lol what. weird it wouldn't be done for both.
